packages/waku/src/lib/plugins/vite-plugin-typegen.ts

```diff
@@ -0,0 +1,102 @@
+import type { Plugin } from 'vite';
+import { promises } from 'fs';
+import path from 'path';
+
+const SRC_DIR = path.resolve(process.cwd(), './src/pages');
+const OUTPUT_FILE = path.resolve(process.cwd(), './entries.gen.tsx');
+
+// Recursively collect `.tsx` files in the given directory
+const collectFiles = async (dir: string): Promise<[string, string][]> => {
+  let results: [string, string][] = [];
+  const files = await promises.readdir(dir, { withFileTypes: true });
+
+  for (const file of files) {
+    const fullPath = path.join(dir, file.name);
+    if (file.isDirectory()) {
+      results = results.concat(await collectFiles(fullPath));
+    } else if (file.isFile() && fullPath.endsWith('.tsx')) {
+      results.push([fullPath, fullPath.replace(SRC_DIR, '')]);
+    }
+  }
+  return results;
+};
+
+const generateFile = async (files: [string, string][]) => {
+  const fileInfo = [];
+  for (const [fullPath, shortPath] of files) {
+    if (shortPath.endsWith('_layout.tsx')) {
+      try {
+        const layoutFile = await import(fullPath);
+        console.log('layoutFile', layoutFile);
+        const config = await layoutFile.getConfig();
+        fileInfo.push({
+          type: 'layout',
+          path: shortPath.replace('_layout.tsx', ''),
+          src: shortPath, // where to import the component from
+          config,
+        });
+      } catch (e) {
+        console.warn('Failed to load layout file:', (e as any).message);
+      }
+    } else if (shortPath.endsWith('index.tsx')) {
+      try {
+        const indexFile = await import(fullPath);
+        const config = await indexFile.getConfig();
+        fileInfo.push({
+          type: 'page',
+          path: shortPath.replace('index.tsx', ''),
+          src: shortPath, // where to import the component from
+          config,
+        });
+      } catch {
+        console.warn('Failed to load index file:', shortPath);
+      }
+    } else {
+      try {
+        const pageFile = await import(fullPath);
+        const config = await pageFile.getConfig();
+        fileInfo.push({
+          type: 'page',
+          path: shortPath.replace('.tsx', ''),
+          src: shortPath, // where to import the component from
+          config,
+        });
+      } catch {
+        console.warn('Failed to load page file:', shortPath);
+      }
+    }
+  }
+  await promises.writeFile(OUTPUT_FILE, JSON.stringify(fileInfo), 'utf-8');
+};
+
+export const typegenPlugin = (): Plugin => {
+  return {
+    name: 'vite-file-watcher-plugin',
+
+    configureServer(server) {
```

Comment:
I noticed that this runs on startup, but it doesn't seem like it runs when I change a file or add/remove/rename files.

I'll probably read through the vite plugin docs more after getting the rest working unless you know a better way to watch for these changes.

Reply:
https://vitejs.dev/guide/api-plugin#handlehotupdate